compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/util/Unifier.java

```diff
@@ -667,6 +667,10 @@ private BType getConstraintTypeIfNotError(BType type) {
             return type;
         }
 
+        if (type instanceof BTypeReferenceType) {
+            return ((BTypedescType) ((BTypeReferenceType) type).referredType).constraint;
```

Comment:
Should we use `Types.getReferredType()` here? If the `TypeReferenceType` is referring to another `TypeReferenceType`, we might get a CCE.

Reply:
Let's make sure to add tests covering these different scenarios too. Given that the Unifier is changed, we probably need dependently-typed function tests also.